storagemarket/provider_loop.go

```diff
@@ -519,6 +528,68 @@ func (p *Provider) run() {
 			}
 			finishedDeal.done <- struct{}{}
 
+		case processedDeal := <-p.processedDealChan:
+			deal := processedDeal.deal
+			sendErrorResp := func(aerr *acceptError) {
+				// If the error is a severe error (eg can't connect to database)
+				if aerr.isSevereError {
+					// Send a rejection message to the client with a reason for rejection
+					resp := acceptDealResp{ri: &api.ProviderDealRejectionInfo{Accepted: false, Reason: aerr.reason}}
+					// Log an error with more details for the provider
+					p.dealLogger.LogError(deal.DealUuid, "error while processing deal acceptance request", aerr)
+					processedDeal.rsp <- resp
+					return
+				}
+
+				// The error is not a severe error, so don't log an error, just
+				// send a message to the client with a rejection reason
+				p.dealLogger.Infow(deal.DealUuid, "deal acceptance request rejected", "reason", aerr.reason, "error", aerr.error)
+				processedDeal.rsp <- acceptDealResp{ri: &api.ProviderDealRejectionInfo{Accepted: false, Reason: aerr.reason}, err: nil}
+			}
+			if processedDeal.err != nil {
+				// Reject offline deal
+				if deal.IsOffline {
+					dh, err := p.mkAndInsertDealHandler(deal.DealUuid)
+					if err != nil {
+						sendErrorResp(&acceptError{error: err, isSevereError: true, reason: "server error: getting deal thread"})
+						continue
+					}
+					dh.close()
+					p.delDealHandler(deal.DealUuid)
+					sendErrorResp(processedDeal.err)
+					continue
+				}
+				// Reject online deal
+				sendErrorResp(processedDeal.err)
+				continue
+			}
+
+			// Accept offline deal
+			if deal.IsOffline {
+				// The deal proposal was successful. Send an Accept response to the client.
+				processedDeal.rsp <- acceptDealResp{ri: &api.ProviderDealRejectionInfo{Accepted: true}}
+				// Don't execute the deal now, wait for data import.
+				continue
+			}
+
+			// Handle online accepted deal
+			// set up deal handler so that clients can subscribe to deal update events
+			dh, err := p.mkAndInsertDealHandler(deal.DealUuid)
+			if err != nil {
+				sendErrorResp(&acceptError{error: err, isSevereError: true, reason: "server error: starting deal thread"})
+				continue
+			}
+
+			// start executing the deal
+			_, err = p.startDealThread(dh, deal)
+			if err != nil {
+				sendErrorResp(&acceptError{error: err, isSevereError: true, reason: "server error: starting deal thread"})
+				continue
+			}
+
+			// send an accept response
+			processedDeal.rsp <- acceptDealResp{&api.ProviderDealRejectionInfo{Accepted: true}, nil}
```

Comment:
* Ditto; see previous comment.
* Can we do something about duplicate code?

Reply:
I have refactored to reduce the duplicate code and fix the RSP channel responses with specific keys.